File: src/main/java/kr/co/bookand/backend/bookstore/repository/BookStoreRepositoryImpl.java
package kr.co.bookand.backend.bookstore.repository;

import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.JPQLQuery;
import com.querydsl.jpa.impl.JPAQueryFactory;
import kr.co.bookand.backend.bookstore.domain.BookStore;
import kr.co.bookand.backend.bookstore.domain.BookStoreTheme;
import kr.co.bookand.backend.bookstore.domain.BookStoreType;
import kr.co.bookand.backend.bookstore.domain.QBookStore;
import kr.co.bookand.backend.common.domain.Status;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.support.PageableExecutionUtils;

import java.util.List;

import static kr.co.bookand.backend.bookstore.domain.QBookStore.*;

@RequiredArgsConstructor
public class BookStoreRepositoryImpl implements BookStoreRepositoryCustom {

    private final JPAQueryFactory queryFactory;

    @Override
    public Page<BookStore> findAllBySearch(String search, String theme, String status, Pageable pageable) {
    JPQLQuery<BookStore> query = queryFactory.selectFrom(bookStore)
                .where(containSearch(search))
                .where(eqTheme(theme))
                .where(eqStatus(status))
                .orderBy(bookStore.id.desc());

        return PageableExecutionUtils.getPage(
                query.offset(pageable.getOffset())
                        .limit(pageable.getPageSize())
                        .fetch(),
                pageable,
                query::fetchCount);
    }

    private BooleanExpression containSearch(String search) {
        if(search == null || search.isEmpty()) {
            return null;
        }
        return bookStore.name.contains(search);
    }

    private BooleanExpression eqTheme(String theme) {
        if(theme == null || theme.isEmpty()) {
            return null;
        }
        BookStoreType bookStoreType = BookStoreType.valueOf(theme);
        return bookStore.themeList.any().theme.eq(bookStoreType);
    }

    private BooleanExpression eqStatus(String status) {
        if(status == null || status.isEmpty()) {
            return null;
        }
        Status statusType = Status.valueOf(status);
        return bookStore.status.eq(statusType);
    }
}

File: src/main/kotlin/kr/co/bookand/backend/bookstore/repository/BookstoreRepositoryImpl.kt
package kr.co.bookand.backend.bookstore.repository

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import kr.co.bookand.backend.account.domain.Account
import kr.co.bookand.backend.bookstore.domain.Bookstore
import kr.co.bookand.backend.bookstore.domain.BookstoreType
import kr.co.bookand.backend.bookstore.domain.QBookstore.*
import kr.co.bookand.backend.common.Status
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.support.PageableExecutionUtils

class BookstoreRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : BookstoreRepositoryCustom {
    override fun findAllBySearch(
        account: Account,
        search: String?,
        theme: String?,
        status: String?,
        pageable: Pageable
    ): Page<Bookstore> {
        val query = queryFactory.selectFrom(bookstore)
            .where(containSearch(search))
            .where(eqTheme(theme))
            .where(eqStatus(status))
            .orderBy(bookstore.id.desc())
        return PageableExecutionUtils.getPage(
            query.offset(pageable.offset)
                .limit(pageable.pageSize.toLong())
                .fetch(),
            pageable,
            query::fetchCount
        )
    }

    private fun containSearch(search: String?): BooleanExpression? {
        if (search.isNullOrEmpty()) {
            return null
        }
        return bookstore.name.contains(search)
    }

    private fun eqTheme(theme: String?): BooleanExpression? {
        if (theme.isNullOrEmpty()) {
            return null
        }
        val bookStoreType = BookstoreType.valueOf(theme)
        return bookstore.themeList.any().theme.eq(bookStoreType)
    }

    private fun eqStatus(status: String?): BooleanExpression? {
        if (status.isNullOrEmpty()) {
            return null
        }
        val statusType = Status.valueOf(status)
        return bookstore.status.eq(statusType)
    }

}